### ISSUE TITLE
Controller version of 'Key'/'Physical Key' split in Input Map

### ISSUE THREAD

**Issue by Sosasees** [godotengine/godot-proposals#4242]
### Describe the project you are working on

A game with controller support

### Describe the problem or limitation you are having in your project

I want to support controllers in my game, but it's unneededly complicated because Godot considers these buttons the same button:

Godot | PlayStation | XBox | Nintendo
------|-------------|------|---------
0     | Cross       | A    | B
1     | Circle      | B    | A
2     | Square      | X    | Y
3     | Triangle    | Y    | X

On both the XBox and Nintendo controllers, the (A) button is normally used as the Accept button and (B) is used for going Back.
But since the buttons are swapped between each controller, the buttons will be correct only on Nintendo or XBox but swapped on the other brand.

This gets even worse with the PlayStation controller, where the button placement is the same worldwide but the buttons are used differently in Japan:

PlayStation | Japan  | America & Europa
------------|--------|-----------------
Cross       | Back   | Accept
Circle      | Accept | Back

(this difference is cultural:
In America and Europe we have the saying 'X marks the spot', which made (Cross) into the Accept button.
But in Japan they use their own versions of our ✅Checkmark and ❎Cross symbols:
they have ⭕Circle instead of ✅Checkmark and ❌Cross instead of ❎Cross, which determined their use of these buttons).

With a big amount of time and effort you can make sure to correctly support all these variations, but most people are lazy like me so they'll just prioritise one controller and accept that the button mapping will be wrong on the other controllers.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

To solve this problem, i propose a split between Physical and Logical Buttons in the Input Map, similar to how it is already done for Keyboard Keys.

Right now we only have Physical Buttons, which is problematic as described above, so we should add Logical Buttons:

|   |        |
|---|--------|
| 0 | Accept |
| 1 | Back   |
| 2 |        |
| 3 |        |

This solves the problem of inconsistent menu navigation between controllers without locking out the possibility of using these face buttons like a second D-Pad for certain types of games.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

In the Input Map's "Plus Button" dropdown, the option "Controller Button" should be renamed to "Physical Controller Button",
so that the new option could be named just "Controller Button".

This new option works as described in the previous section.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This can be wored around with a Lot of lines of code
and buying all the controllers to be able to test that code.
I cannot do that even if i wanted since i have only Nintendo controllers in my household and buying the rest is too expensive right now.

Putting a feature this important in your game shouldn't be this complicated.

### Is there a reason why this should be core and not an add-on in the asset library?

This feature is so impotrant for the players that it should be core.
This is basic accessibility which even normal players need.

**Comment by Mickeon** [comment]
I agree on the problem but I'm not sure about the means, especially if all the necessary Logical Buttons tend to be only two (**Accept** & **Back**). 

Could there be a way to incorporate this in the already existing [JoystickList](https://docs.godotengine.org/en/stable/classes/class_@globalscope.html#JoystickList) constants, as _JOY_ACCEPT_ and _JOY_BACK_? With an inner system similar to [KEY_MASK_CMD](https://docs.godotengine.org/en/stable/classes/class_@globalscope.html#KeyList), but would have to be compared at run-time for the current input device.

**Comment by Sosasees** [comment]
@miv391 this seems like a way better proposal than what i originally had: the same goal but executed much better.
i have just come up with an incremental improvement for it:

A game could be designed to have 2 buttons be pressable with the same thumb (e.g. many Parkour games are designed to have the Run Button be pressable with the same thumb as the Jump Button). To make this kind of control scheme easily possible on all controller types, it should be possible to swap the buttons on the XY row as well, so that we can make sure that the Run Button would always be upper-left next to the Jump Button if we wanted.

**Comment by miv391** [comment]
I think in this case the game programmer should use the physical buttons just as they already work. The right button seems to always be index 1 and the top button is index 3. (I only have a Nintendo Pro Controller and an old Xbox controller so I can't be 100% sure...) So in this case the jump button would be button 1 (B in Xbox and A in Nintendo) and the run button would be button 3 (Y in Xbox and X in Nintendo).

**Comment by Meorge** [comment]
With https://github.com/godotengine/godot/pull/112680, the joypad type (Xbox, PlayStation, Nintendo, etc) will be able to be queried. Perhaps to handle the discrepancy between accept and cancel for Nintendo and other platforms, events in the input map could have a specific scheme (or set of schemes) associated with them, similarly to how a specific device index can be specified?

**Comment by Nintorch** [comment]
See also https://github.com/godotengine/godot/pull/103744

**Comment by poyribn** [comment]
Oh wow, that's a 3&#x2011;years&#x2011;old proposal, that i wouldn't make today because back&nbsp;then, i haven't decided my house&nbsp;style for how face&nbsp;buttons should be mapped for menu&nbsp;navigation in my games:

|   |   |   |
|:-:|:-:|:-:|
| | jump&nbsp;to<br>top&nbsp;menu | |
| jump&nbsp;to<br>left&nbsp;menu | | Accept/<br>Forward |
| | Cancel/<br>Back | |

&#x203b; This mapping is based on a mix of ergonomics and directional&nbsp;metaphors, and is written at the bottom of the game&nbsp;screen, in a strip that's usually named "status&nbsp;bar" in desktop&nbsp;apps.

&#x203b; The use of the face&nbsp;buttons for menu&nbsp;navigation may change per context but i believe the meanings i have written above are good defaults unless playtesting proves otherwise.

But please don't think it means i now dismiss the importance of this proposal.
While i &mdash; the original submitter of this proposal &mdash; no longer need it implemented, i know many other devs would still like to follow platform&nbsp;conventions for the connected controller (not a house&nbsp;style which ignores button&nbsp;labels and uses directional metaphors instead).

**Comment by poyribn** [comment]
API&nbsp;designs that work well for one input&nbsp;method may not work well for other input&nbsp;methods, and this is no exception.

There are so many valid ways that a game&nbsp;designer could reconcile the differences in vendor&#x2011;specific face&nbsp;button layouts, that accounting for all&nbsp;of&nbsp;them in the API could easily be over&#x2011;engineered.

Unlike for keyboards on&nbsp;which the whole layout is different in each language, major differences in gamepad&nbsp;layout that devs may want to treat differently are only on the face&nbsp;buttons.

This means that these functions are enough for devs to reconcile the differences in gamepad&nbsp;layout:

- Get input for both all&nbsp;4 physical face&nbsp;buttons (Up,&nbsp;Right,&nbsp;Down,&nbsp;Left) and the 2 logical face&nbsp;buttons (Accept, Cancel)
[❌ Partially implemented: Only physical layout]
- Query the vendor which invented this face&nbsp;buttons layout:
Microsoft / Nintendo / Sony /
Blank (if face&nbsp;buttons are unlabelled)
[⭕ Implemented, except for the Blank&nbsp;layout which i suspect doesn't exist on any real controller]
- Query the locale to choose the Japanese or Western Sony&nbsp;layout.
Though today, this is only useful to create a deliberately old&#x2011;school control&nbsp;scheme because since the PlayStation&nbsp;5, Sony mandates its Western gamepad&nbsp;layout worldwide.
[⭕ Implemented]
- A Godot&nbsp;documentation page about reconciling differences in vendor&#x2011;specific face&nbsp;buttons layout, which lists all supported layouts (Microsoft, Nintendo, Sony, Blank) and shows the differences between them
[❌ Not written]

We don't need the ability to query inputs for a complete physical gamepad&nbsp;layout and logical gamepad&nbsp;layout separately. We just need to add the 2 logical&nbsp;buttons (Accept,&nbsp;Cancel) to the existing physical gamepad&nbsp;layout for input&nbsp;queries, as that is the only place where the physical and logical gamepad&nbsp;layout may not be the same.
